presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,304 @@
+        private final EqualityInference allInference;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, MultiJoinNode multiJoinNode, CostComparator costComparator)
+        {
+            requireNonNull(idAllocator, "idAllocator is null");
+            requireNonNull(symbolAllocator, "symbolAllocator is null");
+            requireNonNull(session, "session is null");
+            requireNonNull(lookup, "lookup is null");
+            requireNonNull(multiJoinNode, "multiJoinNode is null");
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.session = session;
+            this.planNodeOrdering = getPlanNodeOrdering(costComparator, lookup, session, symbolAllocator);
+            this.allInference = createEqualityInference(multiJoinNode.getFilter());
+        }
+
+        private static Ordering<PlanNode> getPlanNodeOrdering(CostComparator costComparator, Lookup lookup, Session session, SymbolAllocator symbolAllocator)
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+
+        private JoinNode chooseJoinOrder(MultiJoinNode multiJoinNode)
+        {
+            JoinNode join = memo.get(multiJoinNode);
+            if (join == null) {
+                checkState(multiJoinNode.getSources().size() > 1);
+                join = generatePartitions(multiJoinNode.getSources().size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(multiJoinNode, partitioning))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .min(planNodeOrdering)
+                        .orElseGet(() -> createCrossJoins(multiJoinNode));
+                log.debug("Least cost join was: " + join.toString());
+                memo.put(multiJoinNode, join);
+            }
+            return join;
+        }
+
+        /**
+         * This method produces a cross join of all the sources in multiJoinNode.
+         * This gets called once there are no longer edges between any of the join sources.
+         */
+        private JoinNode createCrossJoins(MultiJoinNode multiJoinNode)
+        {
+            return createJoin(
+                    ImmutableSet.of(multiJoinNode.getSources().get(0)),
+                    ImmutableSet.copyOf(multiJoinNode.getSources().subList(1, multiJoinNode.getSources().size())),
```

Comment:
Is that intentional that you join one source with the cross join of all others so that build side is likely bigger than the probe side?
Perhaps it doesn't matter because it will be later flipped in `setJoinNodeProperties`, but...